src/OrleansRuntime/GrainDirectory/LocalGrainDirectory.cs

```diff
@@ -693,77 +758,90 @@ public List<ActivationAddress> GetLocalCacheData(GrainId grain)
-            List<ActivationAddress> addresses = entries.Select(t => ActivationAddress.GetAddress(t.Item1, grain, t.Item2)).ToList();
-            if (log.IsVerbose2) log.Verbose2("FullLookup remote {0}={1}", grain, addresses.ToStrings());
+                // update the cache
+                //List<Tuple<SiloAddress, ActivationId>> entries = result.Item1.Where(t => IsValidSilo(t.Item1)).ToList();
+                //List<ActivationAddress> addresses = entries.Select(t => ActivationAddress.GetAddress(t.Item1, grainId, t.Item2)).ToList();
+                result.Addresses = result.Addresses.Where(t => IsValidSilo(t.Silo)).ToList();
+                if (log.IsVerbose2) log.Verbose2("FullLookup remote {0}={1}", grainId, result.Addresses.ToStrings());
 
-            if (entries.Count > 0)
-                DirectoryCache.AddOrUpdate(grain, entries, result.Item2);
-            
-            return addresses;
+                var entries = result.Addresses.Select(t => Tuple.Create(t.Silo, t.Activation)).ToList();
+
+                if (entries.Count > 0)
+                    DirectoryCache.AddOrUpdate(grainId, entries, result.VersionTag);
+
+                return result;
+            }
         }
 
-        public Task DeleteGrain(GrainId grain)
+        public async Task DeleteGrainAsync(GrainId grainId, int hopCount)
         {
-            SiloAddress silo = CalculateTargetSilo(grain);
-            if (silo == null)
+            // see if the owner is somewhere else (returns null if we are owner)
+            var forwardAddress = this.CheckIfShouldForward(grainId, hopCount, "DeleteGrainAsync");
+
+            // on all silos other than first, we insert a retry delay and recheck owner before forwarding
+            if (hopCount > 0 && forwardAddress != null)
             {
-                // We don't know about any other silos, and we're stopping, so throw
-                throw new InvalidOperationException("Grain directory is stopping");
+                await Task.Delay(RETRY_DELAY);
+                forwardAddress = this.CheckIfShouldForward(grainId, hopCount, "DeleteGrainAsync(recheck)");
             }
 
-            if (log.IsVerbose) log.Verbose("Silo {0} tries to lookup for {1}-->{2} ({3}-->{4})", MyAddress, grain, silo, grain.GetUniformHashCode(), silo.GetConsistentHashCode());
-
-            if (silo.Equals(MyAddress))
+            if (forwardAddress == null)
             {
-                // remove from our partition
-                DirectoryPartition.RemoveGrain(grain);
-                return TaskDone.Done;
+                // we are the owner
+                var registrar = RegistrarManager.Instance.GetRegistrarForGrain(grainId);
+                await registrar.DeleteAsync(grainId);
+            }
+            else
+            {
+                // otherwise, notify the owner
+                DirectoryCache.Remove(grainId);
+                await GetDirectoryReference(forwardAddress).DeleteGrainAsync(grainId, hopCount + 1);
```

Comment:
Was not there in the original, and I do not completely understand it, so I did not add it.